import grpc
import time
import fastf1
import pandas as pd

# ë°©ê¸ˆ ìƒì„±ëœ gRPC ì½”ë“œ ì„í¬íŠ¸
import f1_pb2
import f1_pb2_grpc

# FastF1 ìºì‹œ í´ë” ì„¤ì • (ì—†ìœ¼ë©´ í˜„ì¬ í´ë”ì— ìƒì„±ë¨)
fastf1.Cache.enable_cache('./cache')

def load_f1_data():
    """FastF1ì„ ì´ìš©í•´ 2024ë…„ ëª¬ì GP, ë² ë¥´ìŠ¤íƒ€íœ ë°ì´í„° ë¡œë“œ"""
    print("ğŸï¸  F1 ë°ì´í„°ë¥¼ ë¡œë”© ì¤‘ì…ë‹ˆë‹¤... (ì•½ê°„ì˜ ì‹œê°„ì´ ê±¸ë¦½ë‹ˆë‹¤)")
    session = fastf1.get_session(2024, 'Monza', 'Q')
    session.load()
    
    # ë² ë¥´ìŠ¤íƒ€íœ(VER)ì˜ ê°€ì¥ ë¹ ë¥¸ ë© ì„ íƒ
    lap = session.laps.pick_driver('VER').pick_fastest()
    telemetry = lap.get_telemetry()
    
    # í•„ìš”í•œ ì»¬ëŸ¼ë§Œ ë½‘ê³  NaN ì²˜ë¦¬
    columns = ['Date', 'Speed', 'RPM', 'nGear', 'Throttle', 'Brake', 'DRS','X','Y','Z']
    df = telemetry[columns].fillna(0)
    print(f"âœ… ë°ì´í„° ë¡œë“œ ì™„ë£Œ! ì´ {len(df)}ê°œì˜ ë°ì´í„° í¬ì¸íŠ¸ê°€ ìˆìŠµë‹ˆë‹¤.")
    return df

def generate_telemetry(df):
    """ë°ì´í„°í”„ë ˆì„ì„ ìˆœíšŒí•˜ë©° gRPC ë©”ì‹œì§€ë¥¼ ìƒì„±(Yield)í•˜ëŠ” ì œë„ˆë ˆì´í„°"""
    for _, row in df.iterrows():
        # Proto íŒŒì¼ì— ì •ì˜í•œ TelemetryData ë©”ì‹œì§€ ìƒì„±
        msg = f1_pb2.TelemetryData(
            driver_id="VER",
            timestamp_ms=int(row['Date'].timestamp() * 1000),
            speed=float(row['Speed']),
            rpm=float(row['RPM']),
            gear=int(row['nGear']),
            throttle=float(row['Throttle']),
            brake=float(row['Brake']),
            drs=float(row['DRS']),
            x=float(row['X']),
            y=float(row['Y']),
            z=float(row['Z']),
        )
        
        # ì‹¤ì œ ë ˆì´ìŠ¤ì²˜ëŸ¼ ë³´ì´ê²Œ 0.05ì´ˆ ëŒ€ê¸° (ë¶€í•˜ í…ŒìŠ¤íŠ¸ ì‹œ ì£¼ì„ ì²˜ë¦¬)
        # time.sleep(0.05)
        
        yield msg

def run():
    # Rust ì„œë²„ ì£¼ì†Œ ì—°ê²° (localhost:50051)
    # [::1]ì€ IPv6 localhost, ì•ˆë˜ë©´ '127.0.0.1:50051'ë¡œ ë³€ê²½
    target = 'localhost:50051' 
    
    print(f"ğŸš€ Rust ì„œë²„({target})ë¡œ ì—°ê²° ì‹œë„ ì¤‘...")
    
    with grpc.insecure_channel(target) as channel:
        stub = f1_pb2_grpc.F1TelemetryServiceStub(channel)
        
        try:
            # 1. ë°ì´í„° ë¡œë“œ
            df = load_f1_data()
            
            # 2. ìŠ¤íŠ¸ë¦¬ë° ì „ì†¡ ì‹œì‘
            print("ğŸ“¡ ë°ì´í„° ì „ì†¡ ì‹œì‘!")
            response = stub.StreamTelemetry(generate_telemetry(df))
            
            # 3. ê²°ê³¼ ìˆ˜ì‹ 
            print("ğŸ ì „ì†¡ ì™„ë£Œ!")
            print(f"ì„œë²„ ì‘ë‹µ: {response.status} (ì´ {response.total_packets}ê°œ íŒ¨í‚· ì²˜ë¦¬ë¨)")
            
        except grpc.RpcError as e:
            print(f"âŒ ì—°ê²° ì‹¤íŒ¨: {e}")

if __name__ == '__main__':
    run()